crypto/merkletrie/node.go

```diff
@@ -0,0 +1,383 @@
+			pnode2.leaf = false
+			pnode2.children = make([]storedNodeIdentifier, 256)
+			pnode2.childrenNext = make([]byte, 256)
+			pnode2.children[d[i]] = nodeID
+			pnode2.firstChild = d[i]
+			pnode2.childrenNext[d[i]] = d[i]
+			pnode2.hash = append(path, d[:i]...)
+
+			pnode = pnode2
+			nodeID = nodeID2
+		}
+		return nodeID, nil
+	}
+
+	if n.children[d[0]] == storedNodeIdentifierNull {
+		// no such child.
+		var childNode *node
+		var childNodeID storedNodeIdentifier
+		childNode, childNodeID = cache.allocateNewNode()
+		childNode.leaf = true
+		childNode.hash = d[1:]
+
+		pnode, nodeID = n.duplicate(cache)
+		pnode.children[d[0]] = childNodeID
+
+		if pnode.firstChild > d[0] {
+			pnode.childrenNext[d[0]] = pnode.firstChild
+			pnode.firstChild = d[0]
+		} else {
+			// iterate on all the entries.
+			i := pnode.firstChild
+			for {
+				if pnode.childrenNext[i] < d[0] {
+					if pnode.childrenNext[i] == i {
+						pnode.childrenNext[i] = d[0]
+						pnode.childrenNext[d[0]] = d[0]
+						break
+					}
+				} else {
+					pnode.childrenNext[d[0]] = pnode.childrenNext[i]
+					pnode.childrenNext[i] = d[0]
+					break
+				}
+				i = pnode.childrenNext[i]
+			}
+		}
+	} else {
+		// there is already a child there.
+		childNode, err := cache.getNode(n.children[d[0]])
+		if err != nil {
+			return storedNodeIdentifierNull, err
+		}
+		updatedChild, err := childNode.add(cache, d[1:], append(path, d[0]))
+		if err != nil {
+			return storedNodeIdentifierNull, err
+		}
+		pnode, nodeID = n.duplicate(cache)
+		cache.deleteNode(n.children[d[0]])
+		pnode.children[d[0]] = updatedChild
+	}
+	pnode.hash = path
+	return nodeID, nil
+}
+
+// calculateHash calculate the hash of the non-leaf nodes
+// when this function is called, the hashes of all the child node are expected
+// to have been calculated already. This is achived by doing the following:
+// 1. all node id allocations are done in incremental monolitic order, from the bottom up.
+// 2. hash calculations are being doing in node id incremental ordering
+func (n *node) calculateHash(cache *merkleTrieCache) error {
+	if n.leaf {
+		return nil
+	}
+	path := n.hash
+	hashAccumulator := make([]byte, 0, 64*256)                 // we can have up to 256 elements, so preallocate sufficient storage; append would expand the storage if it won't be enough.
+	hashAccumulator = append(hashAccumulator, byte(len(path))) // we add this string length before the actual string so it could get "decoded"; in practice, it makes a good domain separator.
+	hashAccumulator = append(hashAccumulator, path...)
+	i := n.firstChild
+	for {
+		childNode, err := cache.getNode(n.children[i])
+		if err != nil {
+			return err
+		}
+		lenMod := 1
+		if !childNode.leaf {
+			lenMod = 128
+		}
+		hashAccumulator = append(hashAccumulator, byte(len(childNode.hash)+lenMod)) // we add this string length before the actual string so it could get "decoded"; in practice, it makes a good domain separator.
```

Comment:
Is it possible for `len(childNode.hash)` to be 128 or more? If so this can wrap around.

Reply:
That's true. We might want to target this implementation for "shorter" elements. ( i.e. right now we're using a 36 byte elements. ) having longer element also make the worst case node much deeper.